components/Autocomplete.tsx

```diff
@@ -0,0 +1,184 @@
+import { FC, useState, useEffect, useRef, ReactNode } from 'react'
+
+import {
+  Box,
+  Flex,
+  FlexProps,
+  Input,
+  InputProps as IProps,
+  Popover,
+  PopoverBody,
+  PopoverContent,
+  PopoverTrigger,
+  useDisclosure,
+  useMultiStyleConfig,
+} from '@chakra-ui/react'
+import { TriangleDownIcon } from '@chakra-ui/icons'
+import blem from 'blem'
+import useOutsideClickHandler from 'hooks/useOutsideClickHandler'
+
+const bem = blem('autocomplete')
+
+type OptionType = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  value: any
+  label: string | ReactNode
+  helperText: string | ReactNode
+}
+
+interface AutocompleteProps extends FlexProps {
+  label: string
+  value?: OptionType
+  InputProps?: IProps
+  options?: OptionType[]
+  emptyOption?: ReactNode
+  renderOption?: (option: OptionType) => ReactNode
+  renderSelected?: (option: OptionType) => ReactNode
+  filterOption?: (option: OptionType, searchTerm: string) => boolean
+  onSelectOption?: (option: OptionType) => void
+}
+
+const Autocomplete: FC<AutocompleteProps> = ({
+  label,
+  value = null,
+  InputProps = {},
+  options = [],
+  emptyOption = 'No results',
+  renderOption,
+  renderSelected,
+  onSelectOption = () => void 0,
+  filterOption,
+  ...props
+}) => {
+  const [val, setVal] = useState<OptionType | null>(value)
+  const [search, setSearch] = useState<string>('')
+  const styles = useMultiStyleConfig('Autocomplete', props)
+  const inputRef = useRef<HTMLInputElement>(null)
+  const popoverRef = useRef<HTMLDivElement>(null)
+  const { onOpen, onClose, isOpen } = useDisclosure()
+
+  useOutsideClickHandler([inputRef, popoverRef], onClose)
+  useEffect(() => {
+    setVal(value)
+  }, [value])
+
+  const renderDefaultOption = option => (
+    <Box sx={styles?.option}>
+      <Box className={bem('option-label')}>{option?.label}</Box>
+      <Box className={bem('option-text')}>{option?.helperText}</Box>
+    </Box>
+  )
+  const renderDefaultSelected = option => (
+    <Flex alignItems="end">
+      <Box className={bem('value-label')} pr={2}>
+        {option?.label}
+      </Box>
+      <Box className={bem('value-text')}>{option?.helperText}</Box>
+    </Flex>
+  )
+
+  const optionsToDisplay = options.filter(option => {
+    return filterOption
+      ? filterOption(option, search)
+      : (typeof option?.label === 'string' &&
+          option.label.toLowerCase().includes(search?.toLowerCase())) ||
+          (typeof option?.helperText === 'string' &&
+            option.helperText.toLowerCase().includes(search?.toLowerCase()))
+  })
```

Comment:
Can this be a defaultProps for filterOption function?

Reply:
This function is hard to read. Are the optional chains required? From the type of `OptionType` it doesn't seem like it?

```ts
options.filter((option: OptionType) => {
  const { label: _label, helperText: _helperText } = option
  const label = _label.toLowerCase()
  const help = _helperText.toLowerCase()
  const searchString = search.toLowerCase()
  return filterOption
         ? filterOption(option, search)
         : label && label.includes(search) ||
           help && help.includes(search)
})
```